### PR TITLE
System.Diagnostics Tracing APIs additions

### LINKED ISSUE #40339: Add a way to the ActivityListeners to add more data to the Activity when it gets created
## Background and Motivation

We have exposed the following API in the `ActivitySource` class to create and start the activity.
```C#
    public Activity? StartActivity(string name, ActivityKind kind, ActivityContext parentContext, IEnumerable<KeyValuePair<string, object?>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default)
```

If there is any listener to the Activity events, the parameters passed to `StartActivity` will be sent to the listener to decide if want to create the Activity object and with what initial contents. 
OpenTelemetry samplers create a listener to sample in or out the Activities. We have got a new OpenTelemetry sampler scenario which is, besides deciding to sample in or out, it also needs to add some more tags to the activity when it gets created. That is means we'll need to have a way in our APIs to allow the listeners to pass back some more data we include in the Activity object we'll create later.

Currently, the callback to the listener has the signature:

```C#
    public delegate ActivityDataRequest GetRequestedData<T>(ref ActivityCreationOptions<T> options);
```

Where `ActivityCreationOptions` is a struct that wraps the parameters passed to `StartActivity`. Currently `ActivityCreationOptions` is a read-only struct. The proposal here is to make it read/write struct and adding some extra properties the listeners/sampler can use to stick the extra data (e.g. the extra tags). 
When having `ActivityCreationOptions` as read/write, this will allow us in the future support communicating more data. 

Also, we previously exposed `ActivityListener.AutoGenerateRootContextTraceId` property to allow the listener to communicate if need automatically generating a trace id when we have a root parent. It could be not obvious the purpose of this property when looking at the interfaces. So, we decided to enhance this after having `ActivityCreationOptions` became read/write. The idea is to get rid of the property `AutoGenerateRootContextTraceId` in the listener and then have a TraceId property on the `ActivityCreationOptions`. The property will be a getter only and internally we'll encapsulate the logic of generating a new trace Id when needed.

Last, we are proposing to enhance some of the names used in our APIs.

## Proposed API

```C#
namespace System.Diagnostics
{
    public struct ActivityCreationOptions<T> // We removed the readonly specifier.
    {
        // the getter will automatically create a collection if not created before. 
        // The collection will be used by the listener to add more tags
        public ActivityTagsCollection SamplingTags { get; } 

        // a replacement for ActivityListener.AutoGenerateRootContextTraceId
        public ActivityTraceId TraceId { get; } 
    }
}
```

We are removing the property

```C#
namespace System.Diagnostics
{
    public sealed class ActivityListener : IDisposable
    {
        public bool AutoGenerateRootContextTraceId { get; set;}
    }
}
```

We are renaming the following methods/types:

`ActivityListener.GetRequestedDataUsingParentId` to  `SampleUsingParentId`

`ActivityListener.GetRequestedDataUsingContext` to  `Sample`

The enum `ActivityDataRequest` to `ActivitySamplingResult`

The delegate `GetRequestedData` to `SampleActivity`

Before renaming, we had the following: 

```C#

    public delegate ActivityDataRequest GetRequestedData<T>(ref ActivityCreationOptions<T> options);

    public sealed class ActivityListener : IDisposable
    {
        public GetRequestedData<string>? GetRequestedDataUsingParentId { get; set; }
        public GetRequestedData<ActivityContext>? GetRequestedDataUsingContext { get; set; }
    }
```

After renaming we'll have:

```C#
    public delegate ActivitySamplingResult SampleActivity<T>(ref ActivityCreationOptions<T> options);

    public sealed class ActivityListener : IDisposable
    {
        public SampleActivity<string>? SampleUsingParentId { get; set; }
        public SampleActivity<ActivityContext>? Sample { get; set; }
    }
```

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #40339

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @lmolkova @cijothomas

**Comment by cijothomas** [comment]
@tarekgh the labels are incorrect. it says Globalization.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 1]
```diff
@@ -590,6 +590,211 @@ public void TestEventNotificationOrder()
...
+        public void TestAddSamplerAndActivityCreationTags()
+        {
+            RemoteExecutor.Invoke(() => {
+                using ActivitySource aSource = new ActivitySource("SamplerTags7");
+                using ActivityListener listener = new ActivityListener();
+                listener.ShouldListenTo = (activitySource) => activitySource.Name == "SamplerTags7";
+
+                listener.Sample = (ref ActivityCreationOptions<ActivityContext> activityOptions) =>
+                {
+                    activityOptions.SamplingTags.Add("SamplerTag1", "SamplerValue1");
```
Comment (cijothomas) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:778]: If I set an "tag1" to null in here, will it result in the removal of "tag1" which was provided at the time of startActivity?

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:778]: No, we don't remove any tags from the original set. All tags provided by the sampler will get added regardless the values of teh tags. It is hard to know the intention of the sampler want to add tags with null value or want to remove old tags in the original set.

**Review by noahfalk — APPROVED** [review]
LGTM!

(I suspect there is some room for perf optimization, but given our timing I am not worrying about it for now. I didn't see anything that looked severe)

**Pull request merged**